### DISCUSSION TITLE
Time Optimal Control -> Is there a way to make the 'N' (number of time steps), an Opti variable to be solved.

### DISCUSSION THREAD

**Discussion by subashmu** [casadi/casadi#3328] [Q&A] [answered]
Hello together,

A relevant part of the code from a time optimal control problem using opti stack is below

### Original code
```
opti = casadi.Opti();
X = opti.variable(nx,N+1); % Decision variables for states
U = opti.variable(nu,N); % Decision variables for control vector
T = opti.variable(); % Decision variables for total time

% Gap-closing shooting constraints
for k=1:N
  opti.subject_to(X(:,k+1) == F(X(:,k), U(:,k), T/N));
end
```

I am trying to change the formulation such that the time step 'dT' is fixed already or is made as a parameter. And N is made a variable (Also N should be an integer), so that T = dT * N. So in the new formulation T will go away as below

### Changed code
```
opti = casadi.Opti();
N = opti.variable() % number of time steps
X = opti.variable(nx,N+1); % Decision variables for states
U = opti.variable(nu,N); % Decision variables for control vector
dT = opti.parameter(); % Parameter for time step

% Gap-closing shooting constraints
for k=1:N
  opti.subject_to(X(:,k+1) == F(X(:,k), U(:,k), dT));
end
```
I tried this but it is not working. If some light is thrown on this, it would be very helpful. Also I believe it is a very fundamental problem for the applications where a fixed timestep is already planned; so the answer could help at least a few people now or later.

_Thank you very much for your time!_

**Comment by brunoml5** [comment]
Hi! Take a look at this [reply](https://github.com/casadi/casadi/discussions/3249#discussioncomment-6253354) from Joel. Try reformulation your ODE such that time is another state.

**Reply by subashmu** [reply]
Thanks for your quick response. I have seen the above suggestion of Joel. That is very relevant when the final time tf is unknown. But in my case I'm looking for a way where I want to make 'N' to be a variable and solve for N. At the same time N has to be an integer.

**Reply by brunoml5** [reply]
It is better to avoid integer variables because your problem becomes a Mixed-Integer NLP. Then, you need to use experimental solvers such as BONMIN. Try following the tips from jgillis in the other reply.

**Reply by subashmu** [reply]
Thanks for your suggestions. I'm going to reformulate the problem. Thank you!

**Comment by jgillis** [comment, marked as answer]
Variable dimensions are fundamentally unsupported in CasADi. Perhaps you may mimic this scheme by choosing a numerical Nmax upfront and posing the constraints in such a way that they are relaxed for k>N. Other formulations may be more appropriate. If you are designing a deadbeat controller, perhaps an L1 penalty on tracking error would do the trick.

**Reply by subashmu** [reply]
Thanks for your tips. I'll try to reformulate as you suggested. Thank you!
